doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,895 @@
+==============================================
+NEP 43 — Enhancing the Extensibility of UFuncs
+==============================================
+
+:title: Enhancing the Extensibility of UFuncs
+:Author: Sebastian Berg
+:Status: Draft
+:Type: Standard
+:Created: 2020-06-20
+
+
+.. note::
+
+    This NEP is part of a series of NEPs encompassing first information
+    about the previous dtype implementation and issues with it in
+    :ref:`NEP 40 <NEP40>`.
+    :ref:`NEP 41 <NEP41>` then provides an overview and generic design
+    choices for the refactor. NEPs 42 and 43 (this document) go into the
+    technical details of the internal and external
+    API changes related to datatypes and universal functions, respectively.
+    In some cases it may be necessary to consult the other NEPs for a full
+    picture of the desired changes and why these changes are necessary.
+
+
+Abstract
+--------
+
+Allowing the creation of new user-defined DTypes in NumPy also requires
+a better extensibility of the ufunc interface, which is the backbone for
+a large part of the NumPy API.
+Especially, parametric DTypes are currently severly limited, making it
+impossible for a user-defined DType to register universal function loops.
+
+This NEP proposes a new ``UFuncImpl`` struct, with an extensible C-API
+to register more feature rich loops, thus allowing these use-cases.
+It also proposes to address certain deficiencies, for example with respect
+to error reporting, by extending the signature of the core inner-loop
+functions.
+Finally, one of the most important steps in the ufunc call is finding the
+correct inner loop for a specific set of input dtypes.  We propose to use
+multiple-dispatching in the future.
+
+.. note::
+
+    At this time this NEP is in a preliminary state. Both internal and
+    external API may be adapted based on user input or implementation needs.
+    The general design principles and choices, while provisional, should not
+    be expected to change dramatically.
+    For example, the *exact* additional information passed into ``UFuncImpl``
+    is likely to evolve.
+
+
+Detailed Description
+--------------------
+
+NEP 41 and NEP 42 laid out the general proposed structure into DType
+classes (similar to the current type numbers) and dtype instances.
+Universal functions have a fairly fixed signature of multiple input arrays
+and certain additional features. But depending on the input DTypes, they
+have to perform a different operation.  In general these operations are
+registered by users, e.g. implementers of a user-defined DTypes.
+
+
+Introduction to General Concepts
+""""""""""""""""""""""""""""""""
+
+Universal functions as described above are much like a Python method
+defined on the DType of the array when considering a ufunc with only
+a single input::
+
+    res = np.positive(arr)
+
+could be implemented (conceptionally) as:
+
+    positive_impl = arr.dtype.positive
+    res = positive_impl(arr)
```

Comment:
If this isn't what happens, and not something that the NEP will modify (as you say below), then I'd remove it completely or replace it by a more accurate statement.

Reply:
@hameerabbasi, I am trying to give the reader an idea/concept of what a multi-method is (and how it comes to into play with ufuncs). Starting with normal methods, seemed like a fair way to do that to me?  Maybe the sentence afterwards can say that this is not how it can be done in NumPy for reasons 1 and 2.  (Although, I have only semi appetite to discuss the possibility of an `__array_ufunc__` like approach, unless someone insists on it being listed as a rejected alternative.)

Reply:
How about the following:

```python
# find_actual_implementation is what we want to achieve for custom dtypes
# and ufuncs
positive_impl = find_actual_implementation(np.positive, arr.dtype)
res = positive_impl(arr)
```

I just don't want to give users who gloss over this the perception that `positive` is a `dtype` slot. It's best to choose an example which doesn't require you to explain that.

Reply:
Thanks, I wanted the concept of "method" in there I guess, but maybe that is just not necessary.